pkg/server/delta/v3/server.go

```diff
@@ -44,6 +132,340 @@ func NewServer(ctx context.Context, config cache.ConfigWatcher, callbacks Callba
+	var node = &core.Node{}
+	isWildcard := map[string]bool{}
+
+	for {
+		select {
+		case <-s.ctx.Done():
+			return nil
+			// config watcher can send the requested resources types in any order
+		case resp, more := <-values.deltaEndpoints:
+			if !more {
+				return status.Errorf(codes.Unavailable, "endpoints watch failed")
+			}
+			nonce, err := send(resp, resource.EndpointType)
+			if err != nil {
+				return err
+			}
+			values.deltaEndpointNonce = nonce
+			setState(resp, resource.EndpointType)
+		case resp, more := <-values.deltaClusters:
+			if !more {
+				return status.Errorf(codes.Unavailable, "clusters watch failed")
+			}
+			nonce, err := send(resp, resource.ClusterType)
+			if err != nil {
+				return err
+			}
+			values.deltaClusterNonce = nonce
+			setState(resp, resource.ClusterType)
+		case resp, more := <-values.deltaRoutes:
+			if !more {
+				return status.Errorf(codes.Unavailable, "routes watch failed")
+			}
+			nonce, err := send(resp, resource.RouteType)
+			if err != nil {
+				return err
+			}
+			values.deltaRouteNonce = nonce
+			setState(resp, resource.RouteType)
+		case resp, more := <-values.deltaListeners:
+			if !more {
+				return status.Errorf(codes.Unavailable, "listeners watch failed")
+			}
+			nonce, err := send(resp, resource.ListenerType)
+			if err != nil {
+				return err
+			}
+			values.deltaListenerNonce = nonce
+			setState(resp, resource.ListenerType)
+		case resp, more := <-values.deltaSecrets:
+			if !more {
+				return status.Errorf(codes.Unavailable, "secrets watch failed")
+			}
+			nonce, err := send(resp, resource.SecretType)
+			if err != nil {
+				return err
+			}
+			values.deltaSecretNonce = nonce
+			setState(resp, resource.SecretType)
+		case resp, more := <-values.deltaRuntimes:
+			if !more {
+				return status.Errorf(codes.Unavailable, "runtimes watch failed")
+			}
+			nonce, err := send(resp, resource.RuntimeType)
+			if err != nil {
+				return err
+			}
+			values.deltaRuntimeNonce = nonce
+			setState(resp, resource.RuntimeType)
+		case resp, more := <-values.deltaResponses:
+			if more {
+				if resp == deltaErrorResponse {
+					return status.Errorf(codes.Unavailable, "delta resource watch failed")
+				}
+				typeURL := resp.GetDeltaRequest().TypeUrl
+				nonce, err := send(resp, typeURL)
+				if err != nil {
+					return err
+				}
+				values.deltaNonces[typeURL] = nonce
+				setState(resp, typeURL)
+			}
+		case req, more := <-reqCh:
+			// input stream ended or errored out
+			if !more {
+				return nil
+			}
+			if req == nil {
+				return status.Errorf(codes.Unavailable, "empty request")
+			}
+
+			// Log out our error detail from  envoyif we get one but don't do anything crazy here yet
+			// TODO: embed a logger in the server so we can be a bit more verbose when needed
+			if req.ErrorDetail != nil {
+				return status.Errorf(codes.Code(req.ErrorDetail.GetCode()), "received error from xDS client: %s", req.ErrorDetail.GetMessage())
+			}
```

Comment:
Is this what we do for sotw? Seems bad to not expose this information to the server callbacks, and closing the stream might cause the client to get stuck in a reconnect loop?

Reply:
In SOTW we don't do anything with that errordetail. It just sits and never gets read out which is no good. Exposing that to the callbacks is an interesting though

Reply:
I think logging it is fine (at a low level, it could get super spammy), but terminating the stream seems excessive since there is no reason to believe that restarting the stream will fix it, so it would cause unnecessary churn

Reply:
When I implemented a control plane on top of java-control-plane, we added our own logging based on the server callbacks that would log errors based on this error detail

Reply:
It honestly might be worth exposing another callback that users can implement that would do such a thing. I'll see what I can come up with

Reply:
I think it would make sense to ensure that all requests from the client are made visible through the onRequest callback, that way the implementation of those callbacks is free to do whatever they want with them. Should be easy enough, just pass it to the callback before we do any error handling?